common/src/IO/FreeImageTextureReader.cpp

```diff
@@ -42,37 +42,73 @@ namespace TrenchBroom {
             FIMEMORY*                   imageMemory     = FreeImage_OpenMemory(imageBegin, static_cast<DWORD>(imageSize));
             const FREE_IMAGE_FORMAT     imageFormat     = FreeImage_GetFileTypeFromMemory(imageMemory);
             FIBITMAP*                   image           = FreeImage_LoadFromMemory(imageFormat, imageMemory);
-
             const String                imageName       = path.filename();
+
+            if (image == nullptr) {
+                FreeImage_CloseMemory(imageMemory);
+                return new Assets::Texture(textureName(imageName, path), 64, 64);
+            }
+
             const size_t                imageWidth      = static_cast<size_t>(FreeImage_GetWidth(image));
             const size_t                imageHeight     = static_cast<size_t>(FreeImage_GetHeight(image));
             const FREE_IMAGE_COLOR_TYPE imageColourType = FreeImage_GetColorType(image);
 
-            const auto format = GL_BGR;
+            // This is supposed to indicate whether any pixels are transparent (alpha < 100%)
+            const auto transparent = FreeImage_IsTransparent(image);
+
+            const auto format = GL_RGBA;
             Assets::TextureBuffer::List buffers(m_mipCount);
             Assets::setMipBufferSize(buffers, m_mipCount, imageWidth, imageHeight, format);
 
-            // TODO: Alpha channel seems to be unsupported by the Texture class
-            if (imageColourType != FIC_RGB) {
-                FIBITMAP* tempImage = FreeImage_ConvertTo24Bits(image);
+            const auto inputBytesPerPixel = FreeImage_GetLine(image) / FreeImage_GetWidth(image);
+            if (imageColourType != FIC_RGBALPHA || inputBytesPerPixel != 4) {
+                FIBITMAP* tempImage = FreeImage_ConvertTo32Bits(image);
                 FreeImage_Unload(image);
                 image = tempImage;
             }
 
             FreeImage_FlipVertical(image);
 
-            std::memcpy(buffers[0].ptr(), FreeImage_GetBits(image), buffers[0].size());
-            for (size_t mip = 1; mip < m_mipCount; ++mip) {
+            for (size_t mip = 0; mip < m_mipCount; ++mip) {
                 const auto mipSize = Assets::sizeAtMipLevel(imageWidth, imageHeight, mip);
-                FIBITMAP* mipImage = FreeImage_Rescale(image, static_cast<int>(mipSize.x()), static_cast<int>(mipSize.y()), FILTER_BICUBIC);
-                std::memcpy(buffers[mip].ptr(), FreeImage_GetBits(mipImage), buffers[mip].size());
-                FreeImage_Unload(mipImage);
+                
+                FIBITMAP* mipImage;
+                if (mip > 0) {
+                    mipImage = FreeImage_Rescale(image, static_cast<int>(mipSize.x()), static_cast<int>(mipSize.y()), FILTER_BICUBIC);
+                } else {
+                    mipImage = image;
+                }
+
+                const auto bytesPerPixel = FreeImage_GetLine(mipImage) / FreeImage_GetWidth(mipImage);
```

Comment:
What is going on here? Please document, or even better, extract a function?

Reply:
The bytesPerPixel calculation is from the FreeImage documentation, I'll add a comment and reference the part of the docs I'm following here.

Reply:
or do you mean these for loops below in general? Yeah, I can extract this into a function. It's copying out the contents of the FIBITMAP as a sequence of bytes ordered RGBA, with no padding between the rows. (Internally in the FBITMAP they're stored in an order that is given by the FI_RGBA_* macros, it might not be RGBA, and there is padding between the rows so you can't just memcpy the whole image).

Reply:
Ah, ok. But also I'm wondering why we create mipmaps here when we already have code in `Texture` that will instruct OpenGL to generate them. Is it not better to leave it to GL?

Reply:
Yeah, we could just leave it to OpenGL to create the mipmaps, that would remove the outer loop. This code is also slower because it scales each mipmap from the full size image, instead of scaling it from the next larger mipmap which is what you should do, afaik.

Is it possible to change that without some refactoring to all of the texture loaders? Right now it looks like you pass in the number of mip levels you want to the constructor `FreeImageTextureReader::FreeImageTextureReader`. Maybe we should leave that for later if it requires refactoring all the loaders?

Reply:
I think we could use `FreeImage_ConvertToRawBits` to do this. This function can even invert the image on the fly, so we could ditch that step as well. See the documentation here: http://graphics.stanford.edu/courses/cs148-10-summer/docs/FreeImage3131.pdf

Reply:
We could use that but from my reading of docs/bugs, it doesn't change the byte order, so you still have to check the FI_RGBA_* macros.
https://sourceforge.net/p/freeimage/bugs/172/

Looking at FreeImage.h and FI_RGBA_RED etc., the byte order is either BGRA or RGBA so we could just check for the order and map it to the correct OpenGL enum.

Reply:
`FreeImageTextureReader` is constructed in `TextureLoader::createTextureReader` with a default of 4 mip levels. I think that's fine?

Reply:
Mapping the image format to the correct GL enum seems like the way to go. Less code means less bugs, and also faster texture loading.

Reply:
Agreed, updating it now. Also I see how to skip generating the mips in FreeImageTextureReader.

Reply:
Oh, right, there's more code to generate mips in there. I think that can just be removed. We should leave it to OpenGL to generate them, it'll probably be faster and more correct that way.